### PR TITLE
Overcloud host image improvements

### PR DESCRIPTION
**Pull request by Alex-Welsh**

### CODE REVIEW

**Comment by Alex-Welsh** [comment]
1st commit was reverted so I'll squash them when merging

File: .github/workflows/overcloud-host-image-build.yml [hunk 1]
```diff
@@ -42,6 +52,16 @@ jobs:
           BRANCH=$(awk -F'=' '/defaultbranch/ {print $2}' .gitreview)
           echo "openstack_release=${BRANCH}" | sed "s|stable/||" >> $GITHUB_OUTPUT
 
+      # Generate a tag to apply to all built overcloud host images.
+      - name: Generate overcloud host image tag
+        id: host_image_tag
+        run: |
+          echo "host_image_tag=$(date +${{ steps.openstack_release.outputs.openstack_release }}-%Y%m%dT%H%M%S)" >> $GITHUB_OUTPUT
+
+      - name: Display container image tag
```
Comment (markgoddard) [inline .github/workflows/overcloud-host-image-build.yml]:
```suggestion
      - name: Display host image tag
```

Comment (Alex-Welsh) [reply .github/workflows/overcloud-host-image-build.yml]: Doh. Fixed

**Comment by Alex-Welsh** [comment]
Workflow seems to be working as expected: https://github.com/stackhpc/stackhpc-kayobe-config/actions/runs/4981847393/jobs/8916671378
The run failed because there's no packages in pulp for jammy, but all other images built without issue.

File: .github/workflows/overcloud-host-image-build.yml [hunk 2]
```diff
@@ -98,148 +130,242 @@ jobs:
...
+          --progress
+        env:
+          OS_APPLICATION_CREDENTIAL_ID: ${{ secrets.OS_APPLICATION_CREDENTIAL_ID }}
+          OS_APPLICATION_CREDENTIAL_SECRET: ${{ secrets.OS_APPLICATION_CREDENTIAL_SECRET }}
+        if: inputs.ubuntu-jammy && steps.build_ubuntu_jammy.outcome == 'success' && inputs.sms
 
-      - name: Clean up old images
+      - name: Fail if any host image builds failed
         run: |
-          sudo rm -rf /opt/kayobe/images/
```
Comment (markgoddard) [inline .github/workflows/overcloud-host-image-build.yml:244]: Shouldn't we still clean this up?

File: .github/workflows/overcloud-host-image-build.yml [hunk 3]
```diff
@@ -98,148 +130,242 @@ jobs:
...
-        if: inputs.ubuntu-jammy
+        if: inputs.ubuntu-jammy && steps.build_ubuntu_jammy.outcome == 'success'
 
-      - name: Upload updated images artifact
-        uses: actions/upload-artifact@v3
-        with:
-          name: Updated images list
-          path: /tmp/updated_images.txt
-          retention-days: 7
-        if: always()
```
Comment (markgoddard) [inline .github/workflows/overcloud-host-image-build.yml]: Do you think this is not useful?

File: .github/workflows/overcloud-host-image-build.yml [hunk 4]
```diff
@@ -98,148 +130,242 @@ jobs:
...
+      - name: Fail if any host image builds failed
         run: |
-          sudo rm -rf /opt/kayobe/images/
-        if: always()
+          exit 1
+        if: steps.build_centos_stream_8.outcome == 'failure' ||
+            steps.build_rocky_8.outcome == 'failure' ||
+            steps.build_rocky_9.outcome == 'failure' ||
+            steps.build_ubuntu_focal.outcome == 'failure' ||
+            steps.build_ubuntu_jammy.outcome == 'failure'
```
Comment (markgoddard) [inline .github/workflows/overcloud-host-image-build.yml:438]: While I can see that something failed, it's not currently that clear what it was: https://github.com/stackhpc/stackhpc-kayobe-config/actions/runs/4981847393/jobs/8916671378.

Also, if an image does fail to build we might want to get the build logs out in an artifact.

Comment (markgoddard) [reply .github/workflows/overcloud-host-image-build.yml:438]: hint:
> Image overcloud-ubuntu-jammy failed to build. See overcloud-ubuntu-jammy.stdout and overcloud-ubuntu-jammy.stderr in /opt/kayobe/images/overcloud-ubuntu-jammy.

File: .github/workflows/overcloud-host-image-build.yml (hunk 3 above)
Comment (Alex-Welsh) [reply .github/workflows/overcloud-host-image-build.yml]: I think I removed this when I was working on matrix builds, where I was assuming the workflows might run in parallel on different runners. I'll revert it, but I'll probably also keep the plays that display the updated images during the run.

File: .github/workflows/overcloud-host-image-build.yml (hunk 2 above)
Comment (Alex-Welsh) [reply .github/workflows/overcloud-host-image-build.yml:244]: Yep, fixed

File: .github/workflows/overcloud-host-image-build.yml (hunk 3 above)
Comment (Alex-Welsh) [reply .github/workflows/overcloud-host-image-build.yml]: Done

File: .github/workflows/overcloud-host-image-build.yml (hunk 4 above)
Comment (Alex-Welsh) [reply .github/workflows/overcloud-host-image-build.yml:438]: https://github.com/stackhpc/stackhpc-kayobe-config/actions/runs/4990269078/jobs/8935258654
I've added some logic to return the logs. Workflow is running to test it now

Comment (Alex-Welsh) [reply .github/workflows/overcloud-host-image-build.yml:438]: Looks like it's working, though the stderr artifact was empty which is surprising given it failed

**Comment by Alex-Welsh** [comment]
Ubuntu Jammy builds are now working with the help of the `zstd` on the builder. This PR is now in a reviewable state.

Once this is merged, I will follow up with [this PR](https://github.com/stackhpc/stackhpc-kayobe-config/pull/515) which starts using the images in AIO workflows. It also includes fixes for various issues, allowing those tests to pass.

**Review by markgoddard — APPROVED** [review]

**Pull request merged**